### PR TITLE
Interpreter

### PR DESCRIPTION
**Pull request by jaanonim**
I will write tests later.

### CODE REVIEW

File: core/error.js [hunk 1]
```diff
@@ -82,11 +83,29 @@ class RuntimeError extends CustomError {
         let pos = this.pos_start;
         let ctx = this.context;
 
+        let lastLine = ""
+        let counter = 0;
+
         while (ctx != null) {
-            r += `In ${pos.fn} in ${ctx.displayName} at line ${pos.line + 1}, char: ${pos.column + 1}\n`;
+            let line = `In ${pos.fn} in ${ctx.displayName} at line ${pos.line + 1}, char: ${pos.column + 1}`
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/error.js [hunk 2]
```diff
@@ -82,11 +83,29 @@ class RuntimeError extends CustomError {
         let pos = this.pos_start;
         let ctx = this.context;
 
+        let lastLine = ""
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/error.js [hunk 3]
```diff
@@ -72,6 +72,7 @@ class RuntimeError extends CustomError {
     toString() {
 
         let r = `⚠  ${this.error_name}: ${this.desc}\n`;
+        r += `   For turtle with id ${this.context.symbolTable.get('$who').data}`
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/error.js [hunk 4]
```diff
@@ -82,11 +83,29 @@ class RuntimeError extends CustomError {
...
         while (ctx != null) {
-            r += `In ${pos.fn} in ${ctx.displayName} at line ${pos.line + 1}, char: ${pos.column + 1}\n`;
+            let line = `In ${pos.fn} in ${ctx.displayName} at line ${pos.line + 1}, char: ${pos.column + 1}`
+            if (lastLine == line) {
+                counter++;
+            } else {
+                lastLine = line;
+
+                if (counter > 0) {
+                    r += ` (${counter} times)`
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/error.js [hunk 5]
```diff
@@ -82,11 +83,29 @@ class RuntimeError extends CustomError {
...
+                }
+                r += '\n'
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/error.js [hunk 6]
```diff
@@ -82,11 +83,29 @@ class RuntimeError extends CustomError {
...
+                r += `   ${line}`;
+                counter = 0;
+            }
             pos = ctx.parentEntryPos;
             ctx = ctx.parent;
         }
+        r += '\n'
```
Comment (mwawrzen) [inline core/error.js]: semicolon

File: core/interpreter.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+            new NumberValue(node.token.value)
+            .setPosition(node.pos_start, node.pos_end)
+            .setContext(context)
+        );
+    }
+
+    visitVarNode(node, context) {
+        let res = new RuntimeResult();
+        let var_name = node.token.value;
+        let value = null
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        let res = new RuntimeResult();
+
+        let left = res.register(this.visit(node.left, context));
+        if (res.error) return res;
+        let right = res.register(this.visit(node.right, context));
+        if (res.error) return res;
+
+        let obj = null;
+
+        if (node.token.type === Token.TYPE.PLUS) {
```
Comment (mwawrzen) [inline core/interpreter.js:82]: It can be switch used here

File: core/interpreter.js [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+    visitTellNode(node, context) {
+        let res = new RuntimeResult();
+        let turtles = [];
+        for (let i = 0; i < node.nodes.length; i++) {
+            let value = res.register(this.visit(node.nodes[i], context))
+            if (res.error) return res;
+
+            if (!this.objcts.isTurtle(value.value)) {
+                this.objcts.addTurtle(value.value)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        throw new Error(`No visit method for ${node.constructor.name}`);
+
+        //TODO: Implement this
+    }
+
+    visitTellNode(node, context) {
+        let res = new RuntimeResult();
+        let turtles = [];
+        for (let i = 0; i < node.nodes.length; i++) {
+            let value = res.register(this.visit(node.nodes[i], context))
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        context.symbolTable.setWho(turtles);
+
+        return res.success(null);
+    }
+
+    visitEachNode(node, context) {
+        let res = new RuntimeResult();
+        const ids = context.symbolTable.get("$who").data
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        for (let i = 0; i < ids.length; i++) {
+            context.symbolTable.setWho(ids[i])
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+            res.register(this.visit(node.body, context))
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+            if (res.error) return res;
+        }
+        context.symbolTable.setWho(ids)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+        return res.success(null);
+    }
+
+    visitAskNode(node, context) {
+        let res = new RuntimeResult();
+        const prevIds = context.symbolTable.get("$who").data
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+        for (let i = 0; i < node.nodes.length; i++) {
+            let value = res.register(this.visit(node.nodes[i], context))
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+            if (res.error) return res;
+
+            if (!this.objcts.isTurtle(value.value)) {
+                this.objcts.addTurtle(value.value)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+            }
+
+            context.symbolTable.setWho(value.value)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+            res.register(this.visit(node.body, context))
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+            if (res.error) return res;
+        }
+        context.symbolTable.setWho(prevIds)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        return res.success(null);
+    }
+
+    visitSaveLoadNode(node, context) {
+        let res = new RuntimeResult();
+        if (node.token.isKeyword(Token.KEYWORDS.SAVE)) {
+            let txt = "";
+            const funcs = context.symbolTable.getAllFunc()
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+            for (let i = 0; i < funcs.length; i++) {
+                if (funcs.text !== null) {
+                    txt += funcs[i].text + "\n\n";
+                }
+            }
+            try {
+                fs.writeFileSync(node.path.value, txt);
+            } catch (e) {
+                return res.failure(new RuntimeError(node.pos_start, node.pos_end, `File ${node.path.value} not found! \n     ${e}`, context));
```
Comment (mwawrzen) [inline core/interpreter.js]: weird big white space

File: core/interpreter.js [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+            }
+            return res.success(null)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 24]
```diff
@@ -0,0 +1,273 @@
...
+        } else if (node.token.isKeyword(Token.KEYWORDS.LOAD)) {
+            let txt;
+            try {
+                txt = fs.readFileSync(node.path.value, "utf-8");
+            } catch (e) {
+                return res.failure(new RuntimeError(node.pos_start, node.pos_end, `File ${node.path.value} not found! \n     ${e}`, context));
```
Comment (mwawrzen) [inline core/interpreter.js]: weird big white space

File: core/interpreter.js [hunk 25]
```diff
@@ -0,0 +1,273 @@
...
+            }
+            const Runner = require("./runner");
+            let r = new Runner(node.path.value).run(txt)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

File: core/interpreter.js [hunk 26]
```diff
@@ -0,0 +1,273 @@
...
+            if (r.error) return r;
+            return res.success(null)
```
Comment (mwawrzen) [inline core/interpreter.js]: semicolon

**Review by mwawrzen — APPROVED** [review]
LGTM

File: core/interpreter.js (hunk 8 above)
Comment (jaanonim) [reply core/interpreter.js:82]: I dont use switch

**Review by mwawrzen — APPROVED** [review]
LGTM

**Review by n2oneProgrammer — APPROVED** [review]
LGTM

**Pull request merged**